### ISSUE TITLE
LinearLSQFitter() broken with fixed (or implicit) model terms

### ISSUE THREAD

**Issue by jehturner** [astropy/astropy#6135]
(Moved from #6103:) For some reason when fitting 1D polynomials with only the constant (y intercept) term as a free parameter, the resulting 0-coefficient is offset from its correct value by the mean of the input x co-ordinate array. This does not happen with the other fitters, which give the expected results.
```python
from astropy.modeling.models import *
from astropy.modeling.fitting import *
import numpy as np

x = np.arange(1,6)
y = x+1

fittertype=LinearLSQFitter

model = Linear1D(slope=1)
model.slope.fixed=True
fitter = fittertype()
model_fit = fitter(model, x, y)
print model_fit

# Parameters:
#     slope intercept
#     ----- ---------
#       1.0       3.0

model = Polynomial1D(2, c1=1)
model.c1.fixed=True
fitter = fittertype()
model_fit = fitter(model, x, y)
print model_fit

# Parameters:
#      c0  c1
#     --- ---
#     3.0 1.0
```
The same thing also happens with `Shift()`, which is the real use case here, following some uncommitted changes to make the fitter work with that model.

**Comment by jehturner** [comment]
When there are additional free parameters (but the linear term remains fixed), it produces a different incorrect result:
```python
from astropy.modeling.models import *
from astropy.modeling.fitting import *
import numpy as np

x = np.arange(1,6)
y = x+1

fittertype=LinearLSQFitter

model = Polynomial1D(2, c1=1, c2=0)
model.c1.fixed=True
fitter = fittertype()
model_fit = fitter(model, x, y)
print model_fit

# Parameters:
#           c0       c1       c2      
#     ------------- --- --------------
#     2.23529411765 1.0 0.160427807487
```
But fitting the same model **does** work when unconstrained or when the quadratic term is fixed, so maybe it is specific to having a fixed linear term:
```python
from astropy.modeling.models import *
from astropy.modeling.fitting import *
import numpy as np

x = np.arange(1,6)
y = x+1

fittertype=LinearLSQFitter

model = Polynomial1D(2, c1=1, c2=0)
fitter = fittertype()
model_fit = fitter(model, x, y)
print model_fit

# Parameters:
#      c0  c1         c2        
#     --- --- ------------------
#     1.0 1.0 -1.45744791046e-16

model = Polynomial1D(2, c1=1, c2=0)
model.c2.fixed=True
fitter = fittertype()
model_fit = fitter(model, x, y)
print model_fit

# Parameters:
#      c0  c1  c2
#     --- --- ---
#     1.0 1.0 0.0
```
This does not seem to be a problem with fixing parameters per se, as it also happens when using `Shift()` without any fixed parameters.

Note that the docs say "all fitters support fixed (frozen) parameters ..." and show an example of using `LinearLSQFitter()` with `Polynomial1D()` (http://astropy.readthedocs.io/en/latest/modeling/fitting.html).

**Comment by jehturner** [comment]
I hope to have a quick go at troubleshooting this later in the day (or tomorrow).

**Comment by jehturner** [comment]
Just got back to this... I think this is either:

  1. a profound bug in `LinearLSQFitter()` when dealing with models that are not fully described by their free parameters or
  2. a profound misunderstanding of how fitting is intended to work (in which case the docs could do with spelling this out even more explicitly for dimwits like me).

**Comment by jehturner** [comment]
When `LinearLSQFitter` [passes a matrix equation to `numpy.linalg.lstsq`](https://github.com/astropy/astropy/blob/60dc01d616bd4d30fcfb0666b1b44998466c5e72/astropy/modeling/fitting.py#L330) (let's call it `X.c = Y`), it seems that `Y` contains the _unmodified_ y values supplied by the user, while `X` is the Vandermonde matrix for _only [those coefficients/parameters being fitted](https://github.com/astropy/astropy/blob/60dc01d616bd4d30fcfb0666b1b44998466c5e72/astropy/modeling/fitting.py#L274-L277)_ -- which is insufficient to reproduce the intended model fit. The remaining parameters _do_ get included when the model is later evaluated, but (unlike the non-linear fitters), `LinearLSQFitter` does not actually _use_ model evaluation in its fitting, which therefore happens as if the remaining parameters were zero, producing the wrong results.

Eg. Consider `a + bx` with a fixed `b=1`. What actually gets fitted is just a constant, `a`, by itself, then `b=1` gets added back again afterwards. So if you try to fit `x=[0,1,2,3,4]` and `y=[0,1,2,3,4]`, the constant that best fits `y` is `a=2`, producing a model of `2 + x` instead of the correct `x`.

Shouldn't the unused terms from the Vandermonde matrix (corresponding to fixed parameters) get subtracted from the y values on the right hand side, or something? Note that this must include any fixed terms implicit in the model -- eg. `Shift()` has an implicit linear coefficient of 1 with no corresponding model parameter, which would need to be accounted for somehow.

**Comment by jehturner** [comment]
What the documentation actually says on the matter is this, which is arguably a correct description of incorrect behaviour!

  _For linear fitters, freezing a polynomial coefficient means that a polynomial without that term will be fitted to the data. For example, fixing c0 in a polynomial model will fit a polynomial with the zero-th order term missing. However, the fixed value of the coefficient is used when evaluating the model:_

... but I have trouble seeing how this behaviour could actually be useful or meaningful.

**Comment by jehturner** [comment]
I think the main solution would be to make `_model_to_fit_params()` return the set of parameters _removed_, in addition to the set of parameters to be fitted, then `LinearLSQFitter` could subtract the former from the right-hand side.

The remaining problem is then to deal with any implicitly-fixed terms that don't have associated parameters -- but I _think_ `Shift()` is unique amongst the available linear models in having an implicitly-fixed term that is non-zero -- though it's an important (common) case. I'm not sure what would be the best mechanism for giving the `Shift()` model a fixed term that could get returned as a removed parameter by `_model_to_fit_params()` without the user seeing it as an extra, confusing model parameter. Any ideas?

**Comment by nden** [comment]
@jehturner Sorry, I'm just now catching up with this. I agree with your analysis about the linear fitter and that constant terms need to be subtracted from the free term. Just want to say I will review your PR and will try to get it into v 2.0 if time allows.

**Comment by jehturner** [comment]
Thanks!

**Comment by jehturner** [comment]
The odd thing is that the code did what the documentation said, as if it were intentional, but I have a hard time seeing how the documented behaviour would have been useful to anyone (anyway, the old results could still be reproduced by fixing parameters to 0, fitting and then changing the fixed parameter values).

**Comment by bsipocz** [comment]
This has got fixed by #6174, hasn't it?

**Comment by jehturner** [comment]
Yes.
